File: simulator/Sprite.py
import numpy as np
from .Config import Config
from .Event import Event
from .Events import *

import math
import sys

class Sprite(object):
    def __init__(self, Engine, side, loc, HP, MP, Speed, Armor,
                 ATK, ATKRange, SightRange, Bounty, bountyEXP, BAT, AS):
        self.Engine = Engine
        self.canvas = self.Engine.canvas
        self.v_handle = None
        self.side = side
        self.isDead = False
        self.location = loc
        self.HP = HP
        self.MP = MP
        self.MovementSpeed = Speed
        self.BaseAttackTime = BAT
        self.AttackSpeed = AS
        self.Armor = Armor
        self.Attack = ATK
        self.AttackRange = ATKRange
        self.SightRange = SightRange
        self.Bounty = Bounty
        self.bountyEXP = bountyEXP
        self.LastAttackTime = -1
        self.AttackTime = None
        self.exp = 0

        self.move_target  = None

        self._update_para()
    
    def _update_para(self):
        AttackPerSecond = self.AttackSpeed * 0.01 / self.BaseAttackTime
        self.AttackTime = 1 / AttackPerSecond
    
    def step(self):
        raise NotImplementedError
    
    def draw(self, canvas):
        raise NotImplementedError
    
    def pos_in_wnd(self):
        return (self.location[0] * Config.game2window_scale * 0.5 + Config.windows_size * 0.5,
            self.location[1] * Config.game2window_scale * 0.5 + Config.windows_size * 0.5)
    
    def attack(self,target):
        self.LastAttackTime = self.Engine.get_time()
        AttackEvent.Create(self,target)
    
    def isAttacking(self):
        return self.Engine.get_time() - self.LastAttackTime < self.AttackTime
    
    def set_move(self,target):
        self.move_target = target
    
    def move(self):
        if self.move_target is None:
            return

        #Sprite cant move when it is attacking
        if self.isAttacking():
            return

        dx = self.move_target[0] - self.location[0]
        dy = self.move_target[1] - self.location[1]

        a  = math.atan2(dy,dx)

        if math.isnan(a):
            print("found nan")
            sys.exit(-1)

        if not (self.move_target[0] == 0.0 and self.move_target[1] == 0.0):
            d = self.MovementSpeed * self.Engine.delta_tick
            if math.hypot(dx,dy) < d:
                self.location = self.move_target
            else:
                self.location = (self.location[0] + d *math.cos(a),
                            self.location[1] + d * math.sin(a))
            
        if self.location[0] > Config.bound_length:
            self.location = (Config.bound_length, self.location[1])
        if self.location[1] > Config.bound_length:
            self.location = (self.location[0], Config.bound_length)
        if self.location[0] < -Config.bound_length:
            self.location = (-Config.bound_length, self.location[1])
        if self.location[1] < -Config.bound_length:
            self.location = (self.location[0], -Config.bound_length)
    
    def damadged(self, dmg, dmg_type = None):
        #TODO
        if self.isDead:
            return False
        self.HP -= dmg
        if self.HP <= 0.0:
            print("I'm Dead")
            self.dead()
        return True
    
    def dead(self):
        self.isDead = True
        if self.v_handle != None:
            self.canvas.delete(self.v_handle)
        for s in self.Engine.sprites:
            if Sprite.S2Sdistance(self, s) <= 1300:
                s.exp += self.bountyEXP

    
    @staticmethod
    def S2Sdistance(s1,s2):
        dx = s1.location[0] - s2.location[0]
        dy = s1.location[1] - s2.location[1]
        return math.sqrt(dx * dx + dy * dy)
    

File: cppSimulator/src/Sprite.cpp
#include "Sprite.h"
#include "simulatorImp.h"
#include "Event.h"

#include <cmath>
#include <cstdlib>

void Sprite::attack(Sprite* target)
{
    LastAttackTime = Engine->get_time();
    EventFactory::CreateAttackEvnt(this, target);
}

bool Sprite::isAttacking()
{
    return (Engine->get_time() - LastAttackTime)
        < AttackTime;
}

void Sprite::move()
{
    if (isDead())
        return;

    if (!b_move)
        return;

    if (isAttacking()) {
        b_move = false;
        return;
    }
        

    double dx = std::get<0>(move_target) - std::get<0>(location);
    double dy = std::get<1>(move_target) - std::get<1>(location);
    
    if(dx == 0.0 && dy == 0.0){
        return;
    }

    double a = std::atan2(dy, dx);

    if (std::isnan(a)) {
        printf("found nan dx = %lf dy = %lf\n",dx,dy);
        fflush(stdout);
        exit(3);
    }

    if (!(std::get<0>(move_target) == 0.0
        && std::get<1>(move_target) == 0.0)) {
        double d = MovementSpeed * Engine->get_deltatick();
        if (hypot(dx, dy) < d) {
            location = move_target;
            b_move = false;
        }
        else {
            location = pos_tup(std::get<0>(location) + d * cos(a),
                std::get<1>(location) + d * sin(a));
        }
    }

    if (std::get<0>(location) > Config::bound_length) {
        location = pos_tup(Config::bound_length, std::get<1>(location));
    }
    if (std::get<1>(location) > Config::bound_length) {
        location = pos_tup(std::get<0>(location), Config::bound_length);
    }
    if (std::get<0>(location) < -Config::bound_length) {
        location = pos_tup(-Config::bound_length, std::get<1>(location));
    }
    if (std::get<1>(location) < -Config::bound_length) {
        location = pos_tup(std::get<0>(location), -Config::bound_length);
    }
}

bool Sprite::damadged(Sprite* attacker, double dmg)
{
    if (isDead()) {
        return false;
    }
    HP -= dmg;
    if (HP <= 0.0) {
        dead(attacker);
    }
    return true;
}

void Sprite::dead(Sprite*  attacker)
{
    _isDead = true;
    remove_visual_ent();
    for (Sprite* s : Engine->get_sprites()) {
        if (s->side != side) {
            double dis = S2Sdistance(*s, *this);
            if (dis <= 1300.0) {
                s->exp += bountyEXP;
            }
            //else {
            //    s->exp += bountyEXP * (dis - 1300 + 1) / Config::map_div * 0.1;
            //}
        }
    }
    if (nullptr != attacker) {
        attacker->gold += Bounty;
    }
}

void Sprite::remove_visual_ent()
{
    if (NULL != v_handle) {
        PyObject* delete_fn = PyObject_GetAttrString(canvas, "delete");
        PyObject* args = Py_BuildValue("(O)", v_handle);
        PyObject* kwargs = Py_BuildValue("{}");
        Py_XDECREF(PyObject_Call(delete_fn, args, kwargs));
        Py_DECREF(kwargs);
        Py_DECREF(args);
        Py_DECREF(delete_fn);
        Py_DECREF(v_handle);
        v_handle = NULL;
    }
}

double Sprite::S2Sdistance(const Sprite & s1, const Sprite & s2)
{
    double dx = std::get<0>(s1.location) - std::get<0>(s2.location);
    double dy = std::get<1>(s1.location) - std::get<1>(s2.location);
    return hypot(dx, dy);
}

double Sprite::TimeToDamage(const Sprite * s)
{
    double TimeToAtk = AtkPoint / AttackSpeed * 0.01;
    if (melee == atktype) {
        return TimeToAtk;
    }
    else {
        double _d = S2Sdistance(*this, *s);
        return TimeToAtk + _d / ProjectileSpeed;
    }
}
